core_types/src/json.rs

```diff
@@ -81,9 +98,44 @@ impl<T: Serialize> From<Vec<T>> for JsonString {
     }
 }
 
-impl<T: Serialize, E: Serialize> From<Result<T, E>> for JsonString {
+/// signifies type can be converted to JsonString in Err from some Result
+/// can't use std::error::Error for this because String has Error as a reserved future trait
+pub trait JsonError {}
+
+impl JsonError for HolochainError {}
+
+impl<T: Into<JsonString>, E: Into<JsonString> + JsonError> From<Result<T, E>> for JsonString {
     fn from(result: Result<T, E>) -> JsonString {
-        JsonString::from(serde_json::to_string(&result).expect("could not Jsonify result"))
+        let is_ok = result.is_ok();
+        let inner_json: JsonString = match result {
+            Ok(inner) => inner.into(),
+            Err(inner) => inner.into(),
+        };
+        let inner_string = String::from(inner_json);
+        format!(
+            "{{\"{}\":{}}}",
+            if is_ok { "Ok" } else { "Err" },
```

Comment:
This seems like the right approach, always putting successes and errors behind different keys, but I want to bikeshed a little and propose that the names change, or at least get lowercased. Having a title-cased member name feels awkward to me in any language except Visual Basic, and doesn't even feel Rusty. Maybe the error one could be simply named `"error"`. I'd lean towards `"value"` for the success case.

```javascript
t.ok(result.value)  // assert not an error
t.ok(result.error)  // assert there was an error
```

Reply:
@maackle this is exactly what serde would have done :man_shrugging: 

_every_ enum that we serialize is title case and `Result` is just an enum under the hood, so we get `Ok` and `Err`

while i agree in principle that this gives a somewhat awkward interface, i don't think that the solution is to inconsistently override casing on an ad-hoc basis

can we make this a followup because implementing it properly would need some thought

Reply:
@maackle thinking about this a little more, maybe this logic should go in the node container?

that way the container can decide what is idiomatic for node and lowercase everything across the board. other containers could camel case, kebab case, etc.

for an example of why this goes way beyond `Ok` and `Err` in `Result`, we also have things like this (which i think is much worse) in app spec:

```js
const error = JSON.parse(result.Err.Internal)
```

note `Internal` is also title case and represents a `ZomeApiError::Internal` enum. Rust only serializes the variant name because the compiler always forces you to BYO type, but we don't have that safety/context in JS. If we had another type `FooError::Internal` with the same variant name but different enum, there would be no way of knowing what we meant just from looking at the string.

also note the need to `JSON.parse` that error because it is serialized into a string value due to the way that `hdk::call` works. nested zome calls serialize their internal results each step of the way. In this way `call` (and i assume bridge?) behaves differently to other zome fns that deserialize the results for you on the way out.

and we also have:

```js
const entry_value = JSON.parse(result.Ok.App[1])
```

which (from a JS perspective) is a gross way to work with entries for many reasons.

from the Rust perspective it is great though, it preserves our type safety all the way through JSON, which is the only reason we can get core types at all in zomes! (without heroic effort)

the convo we had the other day about wrapping _everything_ in a giant enum would only make this much worse...

unfortunately all of this seems to just come with the territory of using `serde`. If we want compiler-driven serialization logic then our serialization naturally ends up looking a lot like our compiler types :/

yes we can implement custom serialization logic and serde attributes, but that also seems like a lot of boilerplate/overhead too :(

i already added some items to the SOA around cleaning up the awkwardness in general. we could be putting sugar in the JS side of things that uses contextual knowledge (e.g. that rust/hc is not dynamically typed so returns predictable structures) to do things like `parse_call` and `parse_entry`.

the reason i believe all of this is out of scope for this branch (aside from general PR sizing concerns) is that nothing ^^ actually changed in this branch, all i've done is exposed what was already there :sweat_smile:

Reply:
@thedavidmeister I see, yeah I wasn't thinking about all the possible enums that can serialized and sent as return values. If you think about the entire structure of a zome function return value though, I think it is worth distinguishing the serialization of the root-level `Result` that everything is wrapped in vs. the rest of the enums and structs inside that. Because, if every zome function returns a `Result`, the user has no control over its serialization (unless we do provide config for this on the container side, which does sound like a nice idea), but if they want they can map the other values into more serialize-friendly types before jamming them into the Result on the way out (by using serde `rename` and `tag` attributes on user-defined types).

So I'm not proposing making *all* serialization more idiomatic for other languages, but actually wondering if it makes sense to give special treatment to this top-level Result (which I think it does), since it is indeed being treated specially here even though it's doing what serde would have done.

If others think this should wait for another PR I will shut up about it 😛

Reply:
@maackle it's not true that every zome function returns a `Result` - it is _hdk_ functions that return a `Result`.

zome functions that internally call `hdk::call` actually have serialized internal results - could be an arbitrary level of nesting for this

zome functions that don't use the hdk won't necessarily have a `Result` at all - check out `handle_sum` for example

so the `Result` is not "root level that everything is wrapped in at all" :thinking:

Reply:
another example of when this might be relevant...

we've been getting feedback from the community that they want to do their own serialization that is not JSON at all... so that would look like returning `RawString` from a zome function not a `Result`